sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala

```diff
@@ -882,4 +882,15 @@ class JoinSuite extends QueryTest with SharedSQLContext {
       checkAnswer(df, Row(3, 8, 7, 2) :: Row(3, 8, 4, 2) :: Nil)
     }
   }
+
+  test("SPARK-24495: EnsureRequirements can return wrong plan when reusing the same key in join") {
```

Comment:
as end-to-end test, maybe a better name is: `SPARK-24495: Join may return wrong result when having duplicated equal-join keys`